homeassistant/components/speedtestdotnet/sensor.py

```diff
@@ -98,10 +96,27 @@ def device_state_attributes(self):
 
     async def async_added_to_hass(self):
         """Handle entity which will be added."""
-
-        self.async_on_remove(
-            self.coordinator.async_add_listener(self.async_write_ha_state)
-        )
+        await super().async_added_to_hass()
+        if self.coordinator.config_entry.options[CONF_MANUAL]:
+            state = await self.async_get_last_state()
+            if state:
+                self._state = state.state
+
+        @callback
+        def _update_state():
+            """Update sensors state."""
+            if self.coordinator.data:
+                if self.type == "ping":
+                    self._state = self.coordinator.data["ping"]
+                elif self.type == "download":
+                    self._state = round(self.coordinator.data["download"] / 10 ** 6, 2)
+                elif self.type == "upload":
+                    self._state = round(self.coordinator.data["upload"] / 10 ** 6, 2)
+
+            self.async_write_ha_state()
+
+        self.async_on_remove(self.coordinator.async_add_listener(_update_state))
+        _update_state()
```

Comment:
We shouldn't write state in `async_added_to_hass`. You need to split `_update_state` into 2 methods. Process update and coordinator update (process update + write state).

Also you can make the update state function just class methods.

Reply:
Updated methods as advised. Please check